### PR TITLE
Accessor Library API and Implementation (fixes #97, fixes #95)

### LINKED ISSUE #95: Review Age and Age Unit Indexing Capability
Also explore how to implement multiple age ranges.

┆Issue is synchronized with this [JIRA Story](https://ucsc-cgl.atlassian.net/browse/AZUL-72)
┆Project Name: Azul Indexer
┆Issue Number: AZUL-72

### LINKED ISSUE #97: Extract Indexer code into schema accessor library for use by Orange and Green boxes
Isolate Indexer code from metadata changes by implementing (largely from existing code) and accessor library for the desired facets.

┆Issue is synchronized with this [JIRA Story](https://ucsc-cgl.atlassian.net/browse/AZUL-57)
┆Project Name: azul
┆Issue Number: AZUL-57
┆Epic: Insulate Indexer from metadata

### PR DESCRIPTION
**Pull request by caaespin**
fixes #97
fixes #95

### CODE REVIEW

File: src/azul/dss_bundle.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+from dataclasses import dataclass
+
+
+@dataclass
+class DSSBundle:
+    uuid: str
+    version: str
+    manifest: List[MutableMapping[str, Any]]
+    metadata_files: MutableMapping[str, Any]  # TODO: Change to JSON later
```
[the remark this answers is not in the document]
Comment (caaespin) [reply src/azul/dss_bundle.py]: There is now an EOL at EOF

File: src/azul/project/hca/metadata_api/expando.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+from typing import Any
```
[the remark this answers is not in the document]
Comment (caaespin) [reply src/azul/project/hca/metadata_api/expando.py]: Done

File: src/azul/project/hca/transformers.py [hunk 3]
```diff
@@ -1,17 +1,152 @@
...
-import jmespath
+from typing import Any, List, MutableMapping, Sequence
 
 from azul.transformer import Document, ElasticSearchDocument, Transformer
-from . import extractors
-
-
-# TODO: consider moving the current "create_specimens", etc. to "extract_speciments" and then make a new method called "assign_specimens", etc
-# TODO: Refactor the creation of the subunits.
+from .metadata_api.bundle import (Bundle,
```
[the remark this answers is not in the document]
Comment (caaespin) [reply src/azul/project/hca/transformers.py]: Done

File: test/indexer/test_projects.py [hunk 4]
```diff
@@ -1,11 +1,14 @@
...
+from datetime import datetime
 from typing import Mapping, Any
 from uuid import uuid4
-import logging
 
+from azul.project.hca.config import IndexProperties
+from azul.project.hca.indexer import Indexer
+from azul.project.hca.metadata_api.bundle import Bundle
+
+logging.basicConfig(level=logging.INFO)
```
[the remark this answers is not in the document]
Comment (caaespin) [reply test/indexer/test_projects.py]: Done. See #120 @alex-hancock

File: test/indexer/test_projects.py [hunk 5]
```diff
@@ -105,6 +108,51 @@ def test_same_id_different_bundles(self):
...
+        reconstructed_bundle = Bundle(dss_bundle)
+        sequencing_input = reconstructed_bundle.sequencing_input
+        for si in sequencing_input:
+            print(si.biomaterial_id)
+        specimens = reconstructed_bundle.specimens
+        for sp in specimens:
+            print(sp.biomaterial_id)
+        print(reconstructed_bundle.project)
+
+        self.assertEqual("pass", "pass")
```
[the remark this answers is not in the document]
Comment (caaespin) [reply test/indexer/test_projects.py]: "pass"

File: src/azul/project/hca/metadata_api/bundle.py [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+    def __init__(self, json: ObjectifiedJSON):
+        super().__init__(json)
+        self.instrument_manufacturer_model = json.content.get("instrument_manufacturer_model",
+                                                              {}).get("text")
+
+
+@mydataclass
+class AnalysisProtocol(Protocol):
+
+    def __init__(self, json: ObjectifiedJSON) -> None:
```
[the remark this answers is not in the document]
Comment (caaespin) [reply src/azul/project/hca/metadata_api/bundle.py]: Done

**Comment by caaespin** [comment]
@hannes-ucsc PR is ready for another review

File: src/azul/service/config/request_config.json [hunk 7]
```diff
@@ -5,16 +5,16 @@
...
-    "biologicalSex": "bundles.contents.specimens.sex",
+    "libraryConstructionApproach": "bundles.contents.processes.library_construction_approach",
+    "genusSpecies": "bundles.contents.specimens.genus_species",
+    "organismAge": "bundles.contents.specimens.organism_age",
+    "organismAgeUnit": "bundles.contents.specimens.organism_age_unit",
+    "biologicalSex": "bundles.contents.specimens.biological_sex",
     "disease": "bundles.contents.specimens.disease",
     "instrumentManufacturerModel": "bundles.contents.processes.instrument",
-    "project": "bundles.contents.project.project",
+    "project": "bundles.contents.project.project_shortname",
```
Comment (hannes-ucsc) [inline src/azul/service/config/request_config.json]: These renames should have broken your new tests, @alex-hancock, or required changes to the assertions.

One of the tests, I think the `…correctness` one, should assert the index contents verbatim, field by field.

**Comment by hannes-ucsc** [comment]
I just made those edits and merged.

File: src/azul/project/hca/transformers.py [hunk 8]
```diff
@@ -1,17 +1,143 @@
...
+    def __init__(self) -> None:
+        self.biomaterial_lineage = []
+
+    def visit(self, entity: api.Entity) -> None:
+
+        if isinstance(entity, api.Biomaterial):
+            # noinspection PyProtectedMember
+            b = dict(document_id=entity.document_id,
+                     has_input_biomaterial=entity.has_input_biomaterial,
+                     _source=entity._source)
```
Comment (hannes-ucsc) [inline src/azul/project/hca/transformers.py]: The point about protected fields is that they should not be used outside of a class or its sub-classes.

File: src/azul/project/hca/metadata_api/__init__.py [hunk 9]
```diff
@@ -0,0 +1,598 @@
...
+
+    @classmethod
+    def from_metadata(cls, json: ObjectifiedJSON):
+        described_by = json.content.describedBy.rpartition('/')[2]
+        sub_cls = object_types[described_by]
+        return sub_cls(json)
+
+    def __init__(self, json: ObjectifiedJSON) -> None:
+        super().__init__()
+        self._source = json.content.describedBy.rpartition('/')[2]
```
Comment (hannes-ucsc) [inline src/azul/project/hca/metadata_api/__init__.py]: `source` is an odd name for this, I think. Also don't like the duplication with the line from above.

**Pull request merged**